java/core/src/main/java/com/google/protobuf/DescriptorMessageInfoFactory.java

```diff
@@ -593,21 +616,101 @@ static String getFieldName(FieldDescriptor fd) {
     String name = (fd.getType() == FieldDescriptor.Type.GROUP)
                   ? fd.getMessageType().getName()
                   : fd.getName();
-    String suffix = specialFieldNames.contains(name) ? "__" : "_";
-    return snakeCaseToCamelCase(name) + suffix;
+
+    // convert to UpperCamelCase for comparison to the specialFieldNames
+    // (which are in UpperCamelCase)
+    String upperCamelCaseName = snakeCaseToUpperCamelCase(name);
```

Comment:
I think snakeCaseToUpperCamelCase is only ever used for checking against specialFieldNames. Can we just keep specialFieldNames in lower camel case instead? This would simplify our code and IMO is a bit more intuitive anyways.

Reply:
The specialFieldNames are in UpperCamelCase to handle proto fields named `_my_field`, since `_my_field` will be `MyField` when it is converted to camel case (either upper or lower).  It seemed more straightforward to keep the special field names in UpperCamelCase for comparison (since it handles proto fields named both `my_field`, `_my_field`, and `__my_field`), rather than adding additional logic to handle proto fields starting with `_`. 

Note the tests that use [`ForbiddenWordsLeadingUnderscoreMessage`](https://github.com/protocolbuffers/protobuf/pull/8198/files#diff-c20ad952face0ceea3afab022d736d584b0edf8065f378f052c72a77b0fd3c3fR67) test this case (pun intended ;)).